### PR TITLE
test: Use permissions from git in lint-files.py

### PR DESCRIPTION
**Pull request by laanwj**
Improvements to the `lint-files.py` script:

- Avoid use of `shell=True`.
- Check the permissions in git's metadata instead of in the filesystem. This stops the umask or filesystem from interfering. It's also more efficient as it only needs a single call to `git ls-files`.

(what triggered this change was `File "..." contains a shebang line, but has the file permission 775 instead of the expected executable permission 755.` errors running the script locally).

### CODE REVIEW

File: test/lint/lint-files.py [hunk 1]
```diff
@@ -145,26 +162,26 @@ def check_all_file_permissions() -> int:
             continue
         else:
             print(
```
Comment (laanwj) [inline test/lint/lint-files.py:164]: Could put `assert(0)` here as well, as this is guaranteed to never happen. Git only stores an 'executable' bit, and its expanded permissions will always be 644 or 755.

File: test/lint/lint-files.py [hunk 2]
```diff
@@ -63,17 +70,29 @@ def full_extension(self) -> Optional[str]:
     @property
     def permissions(self) -> int:
         """
-        Returns the octal file permission of the file
+        Returns the octal file permission of the file. Internally, git only
+        keeps an 'executable' bit, so this will always return 0o644 or 0o755.
         """
-        return int(oct(os.stat(self.file_path).st_mode)[-3:])
+        return self._permissions
```
Comment (laanwj) [inline test/lint/lint-files.py:76]: I changed this because I didn't like the octal-as-decimal representation.

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/25015/commits/48d2e80a7479a44b0ab09e87542c8cb7a8f72223

File: test/lint/lint-files.py [hunk 3]
```diff
@@ -178,7 +189,7 @@ def check_shebang_file_permissions() -> int:
                     continue
 
             print(
-                f"""File "{filename}" contains a shebang line, but has the file permission {file_meta.permissions} instead of the expected executable permission {ALLOWED_PERMISSION_EXECUTABLES}. Do "chmod {ALLOWED_PERMISSION_EXECUTABLES} {filename}" (or remove the shebang line)."""
+                    f"""File "{filename}" contains a shebang line, but has the file permission {file_meta.permissions:03o} instead of the expected executable permission {ALLOWED_PERMISSION_EXECUTABLES}. Do "chmod {ALLOWED_PERMISSION_EXECUTABLES} {filename}" (or remove the shebang line)."""
```
Comment (laanwj) [inline test/lint/lint-files.py:192]: I see I accidentally indented here and forgot to add {:03o} for the example permissions. Will fix this on next update.

**Comment by fanquake** [comment]
@vincenzopalazzo want to take another look here?

**Review by vincenzopalazzo — APPROVED** [review]
LGTM!

ACK https://github.com/bitcoin/bitcoin/pull/25015/commits/48d2e80a7479a44b0ab09e87542c8cb7a8f72223

Thanks to ping me @fanquake

**Comment by laanwj** [comment]
> ACK https://github.com/bitcoin/bitcoin/commit/48d2e80a7479a44b0ab09e87542c8cb7a8f72223

Thanks! But that's the old commit :smile:

**Comment by willcl-ark** [comment]
Found this PR after I wasn't able to run test-files.py locally. This patch does indeed fix it.

non-code review ACK.

I'm curious, how exactly does changing to `0o755` fix this? it somehow causes it to respect the default umask as it's now an octal number?

File: test/lint/lint-files.py [hunk 4]
```diff
@@ -11,28 +11,37 @@
...
 
 CMD_TOP_LEVEL = ["git", "rev-parse", "--show-toplevel"]
-CMD_ALL_FILES = "git ls-files -z --full-name"
-CMD_SOURCE_FILES = 'git ls-files -z --full-name -- "*.[cC][pP][pP]" "*.[hH]" "*.[pP][yY]" "*.[sS][hH]"'
-CMD_SHEBANG_FILES = "git grep --full-name --line-number -I '^#!'"
+CMD_ALL_FILES = ["git", "ls-files", "-z", "--full-name", "--stage"]
+CMD_SHEBANG_FILES = ["git", "grep", "--full-name", "--line-number", "-I", "^#!"]
+
+# this intentionally includes uppercase variants, as to be able to flag them
+ALL_SOURCE_FILENAMES_REGEXP = r"^.*\.([cC][pP][pP]|[hH]|[pP][yY]|[sS][hH])$"
```
Comment (MarcoFalke) [inline test/lint/lint-files.py:21]: nit: Could this use `re.IGNORECASE`?

**Comment by laanwj** [comment]
> I'm curious, how exactly does changing to 0o755 fix this? it somehow causes it to respect the default umask as it's now an octal number?

No, that change is an ancillary change (representing octal numbers as decimal is just weird). The main change here is, as noted in the OP, that permissions are queried from git instead of the local filesystem.

File: test/lint/lint-files.py (hunk 4 above)
Comment (laanwj) [reply test/lint/lint-files.py:21]: yes, good idea

**Comment by laanwj** [comment]
Re-pushed [91c78b114ab2e0767e478caf33138147fa1dfff9 → 908fb7e2ec37fe68675d38dbfee4df9f861bb2b5](https://github.com/bitcoin/bitcoin/compare/91c78b114ab2e0767e478caf33138147fa1dfff9..908fb7e2ec37fe68675d38dbfee4df9f861bb2b5)

- Use `re.IGNORECASE` for `ALL_SOURCE_FILENAMES_REGEXP`

**Review by vincenzopalazzo — APPROVED** [review]
Ok LGTM, sorry for the delay and for the error in the commit link

re-tACK https://github.com/bitcoin/bitcoin/commit/908fb7e2ec37fe68675d38dbfee4df9f861bb2b5

**Pull request merged**